Documentation/demo-profile.md

```diff
@@ -0,0 +1,39 @@
+# Inspektor Gadget demo: the "profile" gadget
+
+The profile gadget takes samples of the stack traces.
+Pods can be selected by Kubernetes labels, pod names, namespaces, and nodes.
+Here we deploy a small demo pod "mypod":
+
+```
+$ kubectl run --restart=Never -ti --image=busybox mypod -- sh -c 'for i in $(seq 1 10000) ; do sleep 0.1 ; echo 2*3*7 | bc > /dev/null ; echo Multiplication $i done. ; done'
+```
+
+Using the profile gadget, we can see the list of stack traces.
+The following command filter only for pods named "mypod", execute the command
+and interrupt it after ~30 seconds:
+
+```
+$ ./inspektor-gadget profile --podname mypod
+Node numbers: 0 = ip-10-0-23-61 1 = ip-10-0-3-62^C
+[ 0] sh;[unknown];-;async_page_fault;__do_page_fault;handle_mm_fault;__handle_mm_fault;filemap_map_pages;filemap_map_pages 1
+[ 0] sh;[unknown];[unknown];[unknown];-;async_page_fault;__do_page_fault;up_read;up_read 1
+[ 0] bc;[unknown] 1
+[ 0] sh;[unknown];[unknown];-;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_execve;__do_execve_file.isra.42;do_open_execat;do_filp_open;path_openat;do_dentry_open;ovl_get_inode;cleanup_module;cleanup_module 1
+[ 0] sh;[unknown];[unknown];[unknown];-;async_page_fault;__do_page_fault;handle_mm_fault;__handle_mm_fault;do_wp_page;wp_page_copy;copy_page;copy_page 1
+[ 0] sh;[unknown];[unknown];-;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_execve;__do_execve_file.isra.42;prepare_binprm;security_bprm_set_creds;selinux_bprm_set_creds;security_transition_sid;security_compute_sid.part.17;sidtab_search_core;sidtab_search_core 1
+[ 0] sh;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_exit_group;do_group_exit;do_exit;arch_local_irq_enable;arch_local_irq_enable 1
+[ 0] sh;[unknown] 1
+[ 0] sh;[unknown];[unknown];-;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_execve;__do_execve_file.isra.42;prepare_binprm;security_bprm_set_creds;cap_bprm_set_creds;get_vfs_caps_from_disk;__vfs_getxattr;ovl_xattr_get;vfs_getxattr;security_inode_getsecurity;cap_inode_getsecurity;vfs_getxattr_alloc;ext4_xattr_get;mb_cache_entry_create;mb_cache_entry_create 1
+[ 0] sh;[unknown];-;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_execve;__do_execve_file.isra.42;search_binary_handler;load_elf_binary;flush_old_exec;mmput;exit_mmap;unmap_vmas;unmap_page_range;unmap_page_range 1
+[ 0] sh;[unknown];-;async_page_fault;__do_page_fault;__do_page_fault 1
+[ 0] bc;[unknown];-;async_page_fault;__do_page_fault;__do_page_fault 1
+[ 0] sh;entry_SYSCALL_64_after_hwframe;do_syscall_64;__x64_sys_exit_group;do_group_exit;do_exit;sched_move_task;_raw_spin_unlock_irqrestore;_raw_spin_unlock_irqrestore 1
```

Comment:
To show more interesting stacks, could you try with nginx and wget in a loop?

- `kubectl run nginx --image=nginx`
- Get the pod IP of nginx
- `kubectl run --restart=Never --image=busybox mypod -- sh -c 'while true ; do wget -O /dev/null 10.2.189.5 ; done'`
- `./inspektor-gadget profile --namespace default --label run=mypod`

Does it work?

Reply:
Another simpler example:
```
kubectl run --restart=Never --image=busybox random -- sh -c 'cat /dev/urandom > /dev/null'
```

```
$ ./inspektor-gadget profile --namespace default --podname random
[...]
[ 0] cat;entry_SYSCALL_64_after_hwframe;do_syscall_64;ksys_read;vfs_read;urandom_read;_copy_to_user;copy_user_enhanced_fast_string;copy_user_enhanced_fast_string 17
[ 0] cat;entry_SYSCALL_64_after_hwframe;do_syscall_64;ksys_read;vfs_read;urandom_read;_raw_spin_unlock_irqrestore;_raw_spin_unlock_irqrestore 573
```

But it was executed with this patch:
```
-                               gadgetParams += " -f -d "
+                               gadgetParams += " -f -d -K"
```

So I think we should expose -K and -U first.

Reply:
Are you suggesting to expose flags to the user? If yes, should it be general for all gadgets, right?

About the trace, I'm not sure which one to pick, in all the cases the traces are very long.

Reply:
I was suggesting to only expose `-K` and `-U` in the profile gadget. I don't think exposing all flags without thinking about the use cases would be good.

The traces are sorted by occurrences. The last traces are the most frequent ones, so I would pick the last two traces for the documentation, if the occurrence counter is significantly higher than the previous ones.

Reply:
I was thinking of a generic way to expose flags for other gadgets but I realized that is already done for other gadgets so I followed the same convention.

I exposed the -K and -U flags for profile gadget and updated the documentation to show the example you provided.